keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,479 @@
+    - [GA + 3 cycles](#ga--3-cycles)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, `kube-apiserver` allows one to configure the authorization chain using a set of command line flags of the format `--authorization-*`. And cluster administrators can only enable one webhook as a part of the authorization chain using the `--authorization-modes` flag. This limits admins from creating authorization chains with multiple webhooks that validate requests in a certain order. This proposal makes the case for a more structured config to define the authorization chain for `kube-apiserver` and allowing multiple webhooks with well-defined parameters and enabling fine grained control like explicit Deny authorizer and more definitive `SubjectAccessReviews`.
+
+## Motivation
+
+Cluster administrators should be able to specify more than one authorization webhook in the API Server handler chain. They also need to be able to declaratively configure the authorizer chain using a configuration file.
+
+It should also be easy to say when to Deny requests, for example, when a webhook is unreachable.
+
+### Goals
```

Comment:
in openshift we found a need to allow another authorizer to deny access to a `system:masters` if desired.  We found a need to do this to support a system:master scope limiting his access.

Reply:
We would love to include more user stories!
Could you help us understand the case better?
Do we want to limit the access of role `system:masters` using an additional webhook so that we can have a role system:master with a similar level of access instead?

Reply:
In openshift this is used when a user is a system:master, but wants to produce a scoped token that has narrowed permissions.  An example could be something like scoping a token only operate in a single namespace.  The authorization webhook would be called before the system:masters authorizer so that requests outside the namespace of the scoped token can be rejected.

Reply:
@deads2k how would the `SystemPrivileged` authorizer be enforced when [delegated authorization](https://github.com/kubernetes/kubernetes/blob/1ceca7b139e141014955f4a5307002e33bf9805a/staging/src/k8s.io/apiserver/pkg/server/options/authorization.go#L88-L92) already assumes that `system:masters` can never be denied?

Reply:
Related, how will the RBAC [EscalationAllowed](https://github.com/kubernetes/kubernetes/blob/b60978629d5342b63334c3664737cafcbf84af93/pkg/registry/rbac/escalation_check.go#L38-L44) check work?

Reply:
> This KEP lays the groundwork for being able to mechanically describe that ordering change, but I don't think we should include that feature in this KEP.

I'm willing to defer this